### DISCUSSION TITLE
How do I migrate my data from sqlite to mysql?

### DISCUSSION THREAD

**Discussion by elfensky** [muety/wakapi#728] [Q&A] [answered]
Is this even supported, or am I on my own? 

Given it has a wakatime importer, I thought I'd be able to export my data from the sqlite instance, setup a new mysql instance and import it, but that functionality doesn't seem to exist. 

Perhaps I can convert/export sqlite to mysql?

**Comment by muety** [comment, marked as answer]
While migrating from one database to another is not something that's inherently supported by Wakapi itself, it should still be fairly straightforward. Possible options include:

1. **option: Wakapi import feature** - Basically what you mentioned, i.e. set up a second Wakapi instance, connect it to the first one by API key and then run the importer.
1. **option: CSV import** - Use the [`download_heartbeats`](https://github.com/muety/wakapi/blob/master/scripts/download_heartbeats.py) script to dump your heartbeats to a CSV file, which you can then easily import into the MySQL table (just google for "mysql import csv" or something.
1. **option: Full-database import** - Using some custom script. E.g. [this one](https://stackoverflow.com/a/1067365/3112139) looks promising at first sight.

Note that options 1 and 2 will only copy the raw heartbeats data, but, for example, no users or user settings.

**Reply by elfensky** [reply]
Thank you, I didn't realise the "connect to WakaTime"-feature could also work with wakapi. I was able to import my data with seemingly no issues <3 

For the people in the future:
1. set up a new instance of wakapi with mysql (or whatever database you want). 
2. create a new user and login to the new instance
3. connect your new instance to the old instance
--- url: https://your.wakapi.domain/api/compat/wakatime/v1
--- api: key-for-your-OLD-instance
4. import data
5. ???
6. profit.

**Comment by Dario-Ciceri** [comment]
Hello, I would like to move everything to a new machine on the local network, I have tried on the new machine to enter as the URL http://my-local-ip:3000/api/compat/wakatime/v1 and as the key that of the old instance. When I click on connect I get an error message: Failed to connect to wakatime, api key or endpoint url invalid?

**Reply by elfensky** [reply]
Can you see the logs in the container? Perhaps it does not allow to specify a port and expects it to be the standard 443 (or 80 for http)? 

I recently got a new VPS and set up a fresh wakapi instance using PostGres (my old one used MySQL) and this worked fine, but I did everything over https without specifying a port.

**Reply by Dario-Ciceri** [reply]
Thank you, I exported everything with pgdump and imported it in the new instance before initializing wakapi.

I then used a few commands to change the hash salt (which I found here in the Q&A).

**Reply by muety** [reply]
@elfensky @Dario-Ciceri Thanks for sharing this! 👍

**Reply by da-stoi** [reply]
Couple days too late, but maybe it helps someone else. I had the same issue: 

I had been running Wakapi in a docker container on my local machine for years. I wanted to now migrate to a new instance hosted elsewhere and ran into the same problem. I finally got it to link and import by running everything through a cloudflare tunnel.

`Local Docker Container @ http://my-local-ip:3000 --[Cloudflare Tunnel]--> https://old.waka.domain`

Now from my new Wakapi instance I enter `https://old.waka.domain/api/compat/wakatime/v1` with my api key, and bam, profit!

**Comment by SDS1234** [comment]
@muety, @Dario-Ciceri, @elfensky could you document the steps in the wiki?
I was searching for it but found it now accidentally 😅

**Reply by muety** [reply]
Will do :+1:
